### PR TITLE
Feat: Refactor & Support TP for long-context draft model training

### PR DESCRIPTION
**Pull request by zyksir**
## Motivation

(See #117 ) Training Llama-3.1 models (8B and 70B) in offline mode with long context lengths (e.g., 8K, 16K, or 32K) currently fails with Out-of-Memory (OOM) errors, even on multi-GPU setups.

## Modifications

- Decouple Draft TP and Target DP
- Replicate kv head if `num_kv_head < tp`
- (TODO)Will Merge code from #117 to support TP Feature

Thanks @yd-oom for previous PR

## Related Issues

#112 

## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://sgl-fru7574.slack.com/archives/C09784E3EN6 to discuss your PR.

### CODE REVIEW

File: benchmarks/module_benchmarks/benchmark_flex_attention.py [hunk 1]
```diff
@@ -308,7 +306,7 @@ def plot_results(eagle_results, flex_results, seq_lengths):
     # Define sequence lengths to test
     seq_lengths = [128 * i for i in range(1, 28, 4)]
     # Add extra long context
-    seq_lengths.extend([16384, 32768])
+    seq_lengths.extend([16384])
```
Comment (FrankLeeeee) [inline benchmarks/module_benchmarks/benchmark_flex_attention.py:310]: Why remove 32k?

File: tests/modeling/draft/test_flex_attention.py [hunk 2]
```diff
@@ -19,8 +20,7 @@
     prepare_decoder_attention_mask,
 )
 from specforge.utils import padding
-
-from .utils import norm_tensor
+from tests.utils import norm_tensor
```
Comment (FrankLeeeee) [inline tests/modeling/draft/test_flex_attention.py:23]: This import is not used.

File: specforge/trainer_helper/__init__.py [hunk 3]
```diff
@@ -0,0 +1,4 @@
+from .optimizer import BF16Optimizer
+from .tracker import Tracker, build_tracker
+
+__all__ = ["BF16Optimizer", "Tracker", "build_tracker"]
```
Comment (FrankLeeeee) [inline specforge/trainer_helper/__init__.py:4]: I think this module name is a bit weird, `helper` usually refers to some utilities functions, but this module contains components necessary for training.  I guess these can be an independent file, i.e. `specforge.tracker` and `specforge.optimizer`.

File: specforge/modeling/target/sglang_backend/model_runner.py [hunk 4]
```diff
@@ -64,6 +45,12 @@
 
 
 class SGLangRunner(ModelRunner):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        wrap_eagle3_logits_processors_in_module(self.model)
+
+    def model_specific_adjustment(self):
+        pass
```
Comment (FrankLeeeee) [inline specforge/modeling/target/sglang_backend/model_runner.py:53]: What is this for?

File: specforge/modeling/target/eagle3_target_model.py [hunk 5]
```diff
@@ -182,38 +216,43 @@ def from_pretrained(
...
+                disable_cuda_graph=True,
+                enable_torch_compile=True,
+                tp_size=get_target_tp_size(),
+                pp_size=1,
+                attention_backend="torch_native",
+                mem_fraction_static=0.4,
+            )
+            target_micro_batch_size = None
+        else:
+            server_args = ServerArgs.from_cli_args(args)
```
Comment (FrankLeeeee) [inline specforge/modeling/target/eagle3_target_model.py:237]: I recommend that we don't do this, because this might cause some confusion to the users.
1. Some arguments in SGLang are for optimizations other than prefill, these options won't take effect even if the user specify them
2. this will make `--help` of our training script extremely long.

I recommend that we only keep those important to prefill.

File: scripts/train_eagle3_online.py [hunk 6]
```diff
@@ -47,699 +69,1037 @@
...
 
 
-def parse_args() -> Tuple[ArgumentParser, Namespace]:
-    """
-    This function is used to parse the arguments for the training script.
-    """
-    parser = argparse.ArgumentParser(description="Train Eagle3 with online data")
+@dataclasses.dataclass
+class Eagle3TrainerArgs:
+    target_model_path: str
```
Comment (FrankLeeeee) [inline scripts/train_eagle3_online.py:74]: We can migrate this to `specforge.arguments`.

File: scripts/train_eagle3_online.py [hunk 7]
```diff
@@ -47,699 +69,1037 @@
...
+    target_model_backend: str = "sglang"
+    target_tp_size: int = 1
+    target_dp_size: int = 1
+    target_batch_size: int = -1
+    target_micro_batch_size: int = 8
+    draft_tp_size: int = 1
+    draft_dp_size: int = 1
+    draft_global_batch_size: int = 16
+    draft_micro_batch_size: int = 1
+    draft_accumulation_steps: int = 1
```
Comment (FrankLeeeee) [inline scripts/train_eagle3_online.py:93]: Why do we need micro-batch and globa-batch?

File: benchmarks/module_benchmarks/benchmark_flex_attention.py (hunk 1 above)
Comment (zyksir) [reply benchmarks/module_benchmarks/benchmark_flex_attention.py:310]: 32k will lead to OOM when I test in my H100

File: specforge/modeling/target/sglang_backend/model_runner.py (hunk 4 above)
Comment (zyksir) [reply specforge/modeling/target/sglang_backend/model_runner.py:53]: This function will call something like `log_on_rank0`; here, dist is initialized but the sglang parallel_state._WORLD is not set, which will lead to an error
